bak2dvd/bak2dvd

```diff
@@ -0,0 +1,268 @@
+  fi
+}
+
+checkInternet()
+{
+  httpGet github.com > /dev/null 2>&1 || { echo "Error: no active internet connection" >&2; return 1; } # query github with a get request
+}
+
+usage()
+{
+  cat <<EOF
+Bak2dvd
+Description: A script to write tar archives to optical disc.
+Usage: bak2dvd [flags] -t "[arguments]"
+  -u  Update Bash-Snippet Tools
+  -h  Show the help
+  -v  Get the tool version
+  -t  Parameters to pass to tar
+
+ Disc technology flags are:
+
+  -cd       for cd-rom"
+  -dvd      for single layer DVD"
+  -dvd-dl   for dual layer DVD"
+  -bd       for single layer BluRay"
+  -bd-dl    for dual layer BluRay"
+
+ Preset tar parameters are -cvpML and -f
+
+Example:
+   bak2dvd -d /dev/sr0 -dvd -t "-C $HOME Documents"
+   tool
+EOF
+}
+
+
+options=0
+
+while test $# -gt 0; do
+  case "$1" in
+    \?) echo "Invalid option: -$1" >&2
+        exit 1
+        ;;
+   -h|--help)  usage
+        exit 0
+        ;;
+    -v)  echo "Version $currentVersion"
+        exit 0
+        ;;
+    -u)  getConfiguredClient || exit 1
+        checkInternet || exit 1
+        update
+        exit 0
+        ;;
+    -d)
+	shift
+	if test $# -gt 0; then
+		export device=$1
+	else
+		echo "No optical device specified"
+		exit 1
+	fi
+	shift
+	export options=1
+	;;
+     -dvd)
+		export tracklen="4580000" # DVD+R capacity - 10M
+		shift
+		export options=1
+		;;
+     -dvd-dl)
+		export trackle71n="8337408" # DVD+R DL capacity - 10M
+		shift
+		export options=1
+		;;
+     -bd)
+		export tracklen="24428544" # This should be 10M less than total
+		shift
+		export options=1
+		;;
+     -bd-dl)
+		export tracklen="48867328" #same as above but dual layer
+		shift
+		export options=1
+		;;
+     -cd)
+		export tracklen="712702" #700M cd-rom capacity - 4k, 10M is not an insigficant amout at this storage capacity!
+		shift
+		export options=1
+		;;
+     -t)
+		shift
+		if test $# -gt 0; then
+			export taropts=$1
+		else
+			echo "No options to tar specified."
+			exit 1
+		fi
+		shift
+		export options=1
+		;;
+    :)  echo "Option -$OPTARG requires an argument." >&2
+        exit 1
+        ;;
+  esac
+done
+
+# special set of first arguments that have a specific behavior across tools
+if [[ $# == "0" && $options == "0" ]]; then
+  usage ## if calling the tool with no flags and args chances are you want to return usage
+  exit 0
+elif [[ $# == "1" ]]; then
+  if [[ $1 == "update" ]]; then
+    getConfiguredClient || exit 1
+    checkInternet || exit 1
+    update || exit 1
+    exit 0
+  elif [[ $1 == "help" ]]; then
+    usage
+    exit 0
+  fi
+fi
+
+## The rest of the conditions and code would go here
+## Make sure to use checkInternet at least once before any time httpGet will be called.
+## Make sure to call getConfiguredClient at least once before ever calling checkInternet.
+
+pipename="/tmp/$RANDOM.pipe"
+fifoname="/tmp/$RANDOM.fifo"
+mknod $pipename p
+mkfifo $fifoname
+# Asign the fifo to a file descriptor. No need for ugly cat hack, and it actually works!
```

Comment:
Can you explain whats going on here in the next line, looks interesting

Reply:
This is creating a pipe to feed the burning programs from tar. Tar when it's requesting the user to prepare the next volume needs user input in the form of a \n to continue. In this script tar can't receive that from the user because it's been sent to the background. A named pipe can be used to communicate to a process in the background, but it needs to be kept open somehow. One way that seems to work for other people is cat >pipe.fifo &. I consider this an ugly hack compared with using exec <>7 pipe.fifo to open the pipe in a bidirectional mode which keeps it from blocking. Also cat failed me by  dieing right after tar launched causing it to receive an EOF when it was requesting input from the user. Tar apparently doesn't like this and terminates with an error message. So this made the trick with cat an ugly hack in my eyes when exec can be used in a way to open a fifo and not need to clean a dummy processes at the same time.